### ISSUE TITLE
Support for default token value in route definition syntax

### ISSUE THREAD

**Issue by Kallys** [bcosca/fatfree-core#190]
Hi !

I extended route definition syntax with default value for token, in one of my project.
I didn't really take time to see how routing engine works, so I did a abstraction layer over it.

It allow me to use syntax like this :
`/my/page/@page_id/@action="show"/@user_id="22"/@?admin_key=""`

Which route like this :
```php
['page_id' => 10] // -> "/my/page/10/show/22"
['page_id' => 10, 'action' => 'remove'] // -> "/my/page/10/remove/22"
['page_id' => 10, 'user_id' => 33] // -> "/my/page/10/show/33"
['page_id' => 10, 'action' => null, admin_key => 'h@$h'] // -> "/my/page/10/show/22/h@$h"
```

With a restriction for optional tokens (syntax "@?") : they are allowed only if they precede another optional token, or if they are ending route definition.
```php
/my/page/@page_id/@?action="show"/@?user_id="22"/@?admin_key="" // legal
/my/page/@page_id/@?action="show"/@user_id="22"/@?admin_key="" // illegal
/my/page/@page_id/@?action="show"/@user_id="22"/@admin_key="" // illegal
/my/page/@page_id/@action="show"/@user_id="22"/@?admin_key=""/success // illegal
```

To make this working, I use an hive value which contains default values and order of each token per alias, and then generate all possible tokenized routes for F3's Base routing system.

I found it really handy and wanted to share the idea. Unfortunately, project (and its open source code) using this feature should be release in 2018... But maybe, you may want to add this feature directly inside routing engine !

Thanks !

**Comment by xfra35** [comment]
Hi,

Does your code patch the [build](https://github.com/bcosca/fatfree-core/blob/f25c17b9b00fe594e6ea0ece5d9952fc580844cb/base.php#L161) function as well?

**Comment by Kallys** [comment]
No, I did not want to patch F3 code since I want to avoid as much as possible to maintain a patch every times F3 releases an update.
So my code uses a higher abstraction layer, an external class computing routes using F3 build function inside. Basically, my class purpose is just to choose right tokens to send to F3 build function.
Then, I use a template filter to use my router in F3 template language (or call directly my class in code).
